### PR TITLE
add fr tour singleton objects

### PR DESCRIPTION
**Pull request by apointeau**

### CODE REVIEW

File: _fr/tour/singleton-objects.md [hunk 1]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
-num: 12
+num: 15
 
 language: fr
 
 next-page: regular-expression-patterns
 previous-page: pattern-matching
 ---
+
+Un objet est une classe qui a exactement une instance. Il est créé de façon passive au moment où il est référencé, comme une valeur passive `lazy val`.
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:14]:
```suggestion
Un objet est une classe qui a exactement une instance. Il est créé de façon paresseuse au moment où il est référencé, comme une valeur paresseuse `lazy val`.
```

File: _fr/tour/singleton-objects.md [hunk 2]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+
+En tant que valeur de premier niveau, un object est un singleton.
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:16]:
```suggestion
En tant que valeur de premier niveau, un objet est un singleton.
```

File: _fr/tour/singleton-objects.md [hunk 3]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+
+# Définir un objet singleton
+
+Un objet est une valeur. La définition d'un objet ressemble a une classe, mais utilise le mot clef `object` :
+
+```scala mdoc
+object Box
+```
+
+Voici l'exemple d'un objet avec une méthode :
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:28]:
```suggestion
Voici un exemple d'un objet avec une méthode :
```

File: _fr/tour/singleton-objects.md [hunk 4]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+class Test {
+  val project1 = new Project("TPS Reports", 1)
+  val project2 = new Project("Website redesign", 5)
+  info("Created projects")  // Prints "INFO: Created projects"
+}
+```
+
+La méthode `info` est visible grâce à l'import, `import logging.Logger.info`. Les imports ont besoin d'un chemin d'accès stable aux ressources, et un objet est un chemin stable.
+
+Note : Si un `objet` est encapsulé dans une autre classe ou un autre objet, alors l'objet est dépendant du chemin d'accès comme les autres membres. Cela veut dire par exemple, que si on prend 2 types de boissons, `class Milk` et `class OrangeJuice`, un membre de class `object NutritionInfo` est dépendant de son instance d'encapsulation. `milk.NutritionInfo` est complétement différent de `oj.NutritionInfo`.
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:56]:
```suggestion
Note : Si un `objet` est encapsulé dans une autre classe ou un autre objet, alors l'objet est dépendant du chemin d'accès, comme les autres membres. Cela veut dire, par exemple, que si on prend 2 types de boissons, `class Milk` et `class OrangeJuice`, un membre de classe `object NutritionInfo` est dépendant de son instance d'encapsulation. `milk.NutritionInfo` est complètement différent de `oj.NutritionInfo`.
```

File: _fr/tour/singleton-objects.md [hunk 5]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+
+## Les objets compagnons
+
+Un objet avec le même nom qu'une classe est appelé un _objet compagnon_. Inversement, la classe se nomme la _classe compagnon_ de l'objet. Une classe ou un objet compagnon peut accéder aux membres privés de son compagnon.
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:60]: This sentence from the original does not appear to have been translated:

> Use a companion object for methods and values which are not specific to instances of the companion class.

File: _fr/tour/singleton-objects.md [hunk 6]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+       |Username: ${email.username}
+       |Domain name: ${email.domainName}
+     """.stripMargin)
+  case None => println("Error: could not parse email")
+}
+```
+
+L'objet `object Email` contient une méthode de fabrique `fromString` qui créé une instance de `Email` depuis une chaîne de caractères. L'instance est retournée en tant que `Option[Email]` pour gérer le cas des erreurs de syntaxe.
+
+Note : Si une classe ou un objet a un compagnon, tout deux doivent être définis dans le même fichier. Pour définir des compagnons dans le REPL, tout deux doivent être définis sur la même ligne ou il faut entrer en mode `:paste`.
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:108]:
```suggestion
Note : Si une classe ou un objet a un compagnon, tous deux doivent être définis dans le même fichier. Pour définir des compagnons dans le REPL, tous deux doivent être définis sur la même ligne ou définis en mode `:paste`.
```

File: _fr/tour/singleton-objects.md [hunk 7]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+     """.stripMargin)
+  case None => println("Error: could not parse email")
+}
+```
+
+L'objet `object Email` contient une méthode de fabrique `fromString` qui créé une instance de `Email` depuis une chaîne de caractères. L'instance est retournée en tant que `Option[Email]` pour gérer le cas des erreurs de syntaxe.
+
+Note : Si une classe ou un objet a un compagnon, tout deux doivent être définis dans le même fichier. Pour définir des compagnons dans le REPL, tout deux doivent être définis sur la même ligne ou il faut entrer en mode `:paste`. 
+
+## Notes pour les programmeurs Java ##
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:110]:
```suggestion
## Notes pour les programmeurs Java
```

File: _fr/tour/singleton-objects.md [hunk 8]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+
+Les membres `static` en Java sont modélisés comme des membres ordinaires d'un objet compagnon en Scala.
+
+En utilisant un objet compagnon depuis du code Java, ses membres seront définis dans la classe compagnon avec le modificateur `static`. Cela s'appelle le _static forwarding_. Cela se produit même si vous n'avez pas défini de classe compagnon vous-même.
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:114]:
```suggestion
Lorsqu'on utilise un objet compagnon depuis du code Java, ses membres sont définis dans la classe compagnon avec le modificateur `static`. Cela s'appelle le _static forwarding_. Cela se produit même si vous n'avez pas défini de classe compagnon vous-même.
```

File: _fr/tour/singleton-objects.md [hunk 9]
```diff
@@ -3,10 +3,118 @@ layout: tour
...
+
+## Plus d'informations
+
+* Apprennez-en plus sur les objets compagnons dans [Scala Book](/overviews/scala-book/companion-objects.html)
```
Comment (sjrd) [inline _fr/tour/singleton-objects.md:118]:
```suggestion
* Apprenez-en plus sur les objets compagnons dans le [Scala Book](/overviews/scala-book/companion-objects.html)
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Thanks for the PR. I have a few comments below.

File: _fr/tour/singleton-objects.md (hunk 5 above)
Comment (apointeau) [reply _fr/tour/singleton-objects.md]: I knew that I missed a sentence, but I was thinking that I found it. Ty, I fixed it in the last commit with :

> L'objet compagnon est utile pour les méthodes et les valeurs qui ne sont pas spécifiques aux instances de la classe compagnon.

**Review by sjrd — APPROVED** [review]

**Pull request merged**